lib/internal/fs/recursive_watch.js

```diff
@@ -0,0 +1,282 @@
+
+    files.set(f, file);
+
+    if (file.isDirectory()) {
+      ArrayPrototypePush(subdirectories, traverse(f, files));
+    }
+  }
+
+  await SafePromiseAll(subdirectories);
+
+  return files;
+}
+
+class FSWatcher extends EventEmitter {
+  #options = null;
+  #closed = false;
+  #files = new SafeMap();
+  #rootPath = path.resolve();
+  #watchingFile = false;
+
+  constructor(options = kEmptyObject) {
+    super();
+
+    assert(typeof options === 'object');
+
+    const { persistent, recursive, signal, encoding } = options;
+
+    // TODO(anonrig): Add non-recursive support to linux-watcher for IBMi support.
+    if (recursive != null) {
+      validateBoolean(recursive, 'options.recursive');
+    }
+
+    if (persistent != null) {
+      validateBoolean(persistent, 'options.persistent');
+    }
+
+    if (signal != null) {
+      validateAbortSignal(signal, 'options.signal');
+    }
+
+    if (encoding != null) {
+      // This is required since on macOS and Windows it throws ERR_INVALID_ARG_VALUE
+      if (typeof encoding !== 'string') {
+        throw new ERR_INVALID_ARG_VALUE(encoding, 'options.encoding');
+      }
+    }
+
+
+    this.#options = { persistent, recursive, signal, encoding };
+  }
+
+  close() {
+    if (this.#closed) {
+      return;
+    }
+
+    const { unwatchFile } = lazyLoadFsSync();
+    this.#closed = true;
+
+    for (const file of this.#files.keys()) {
+      unwatchFile(file);
+    }
+
+    this.#files.clear();
+    this.emit('close');
+  }
+
+  /**
+   * @param {string} file
+   * @return {string} the full path of `file` if we're watching a directory, or the
+                      filename when watching a file.
+   */
+  #getPath(file) {
+    // Always return the filename instead of the full path if watching a file (not a directory)
+    if (this.#watchingFile) {
+      return StringPrototypeSlice(this.#rootPath, StringPrototypeLastIndexOf(this.#rootPath, '/') + 1);
+    }
+
+    // When watching directories, and root directory is changed, return the full file path
+    if (file === this.#rootPath) {
+      return this.#rootPath;
+    }
+
+    // Otherwise return relative path for all files & folders
+    return path.relative(this.#rootPath, file);
+  }
+
+  #unwatchFolder(file) {
+    const { unwatchFile } = lazyLoadFsSync();
+
+    for (const filename of this.#files.keys()) {
+      if (StringPrototypeStartsWith(filename, file)) {
+        unwatchFile(filename);
+      }
+    }
+  }
+
+  async #watchFolder(folder) {
+    const { opendir } = lazyLoadFsPromises();
+
+    try {
+      const files = await opendir(folder);
+
+      for await (const file of files) {
+        if (this.#closed) {
+          break;
+        }
+
+        const f = path.join(folder, file.name);
+
+        if (!this.#files.has(f)) {
+          this.#files.set(f, file);
+          this.emit('change', 'rename', this.#getPath(f));
+
+          if (file.isDirectory()) {
+            await this.#watchFolder(f);
```

Comment:
I worry about this await here in terms of performance. Does this *need* to await or can this some how be queued to operate in parallel with the outer loop?

Also, we should be careful about symlinks here that set up an infinite recursion. I know it's an edge case but this should likely try to account for it.

Simple test case:

```
> mkdir a
> cd a
> ln -s . b
```

Reply:
Should we skip symlinks, or just include them and handle the infinite recursion case?

Reply:
What happens if a symlink exists, and the symlink is a folder outside of the parent folder? Should we recursively follow the symlink too?

Reply:
Good questions. I don't know. Skipping symlinks feels wrong, tho.